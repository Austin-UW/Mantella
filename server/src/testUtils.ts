import { ProjectFieldsFragment } from './types'
import { MutationResult } from 'react-apollo'
import { ApolloServer } from 'apollo-server-express'
import { queries } from './graphql/queries/queries'
import { mutations } from './graphql/mutations/mutations'
import path from 'path'
import fs from 'fs'
import mongoose from 'mongoose'

// Load models since we will not be instantiating our express server.
mongoose.connect(
  process.env.DB_CONNECT ||
    'mongodb+srv://conrad:cokay101@kanbanbrawn-41gbz.mongodb.net/kanban?retryWrites=true',
  { useNewUrlParser: true }
)

mongoose.set('useFindAndModify', false)
mongoose.set('useCreateIndex', true)

const { createTestClient } = require('apollo-server-testing')

const resolvers: any = {
  Query: queries,
  Mutation: mutations
}

const typeDefs: any = [
  fs.readFileSync(path.join(__dirname, './graphql/schema.graphql'), 'utf8')
]

const gqlServer = new ApolloServer({
  typeDefs,
  resolvers,
  tracing: false,
  context: ({ req, res }) => {
    return {
      req,
      res,
      user: {
        projects: ['5d018bc58104190fe6e2ac2c'],
        _id: '5d018bc58104190fe6e2ac2d',
        password:
          '$2a$10$LL5wFKVW./hWctu8lypf/.geDiezQhQG6rYdbI7fIsVd0wfzYJwha',
        email: 'conradkay123@gmjail.com',
        username: 'kachow',
        profileImg:
          'https://mb.cision.com/Public/12278/2797280/879bd164c711a736_800x800ar.png',
        __v: 0
      }
    }
  }
})

export const test = createTestClient(gqlServer)

export const gqlReq = async <Q, V>(args: {
  query: any
  variables: V
  isMutation: boolean
  onError?: (err: any) => void
}): Promise<MutationResult<Q>> => {
  const result = args.isMutation
    ? await test.mutate({ mutation: args.query, variables: args.variables })
    : await test.query({ query: args.query, variables: args.variables })

  return result
}

export const projToSnapshot = (project: ProjectFieldsFragment) => {
  return {
    ...project,
    id: 'PROJECT_ID',
    columns: project.columns.map((column, i) => {
      return {
        ...column,
        id: `COL_ID ${i}`,
        taskIds: column.taskIds.map((taskId, e) => `TASK_ID ${e}`)
      }
    }),
    columnIds: project.columnOrder.map((columnId, i) => `COL_ID ${i}`),
    ownerId: 'OWNER_ID',
    swimlanes: project.swimlanes.map((swimlane, i) => {
      return {
        ...swimlane,
        id: `SWIM_ID ${i}`,
        taskIds: swimlane.taskIds.map((taskId, e) => `TASK_ID ${e}`)
      }
    }),
    tags: project.tags!.map((tag, i) => {
      return { ...tag, id: `TAG_ID ${i}` }
    }),
    tasks: project.tasks.map((task, i) => {
      return {
        ...task,
        id: `TASK_ID ${i}`,
        subTasks: task.subTasks.map((subTask, e) => {
          return { ...subTask, id: `SUBTASK_ID ${e}` }
        }),
        comments: task.comments.map((comment, e) => {
          return { ...comment, id: `COMMENT_ID ${e}` }
        })
      }
    }),
    users: project.users!.map((user, i) => {
      return { id: `USER_ID ${i}` }
    })
  }
}
