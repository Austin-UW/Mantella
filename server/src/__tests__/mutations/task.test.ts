import {
  CreateTaskMutation,
  CreateTaskMutationVariables,
  EditTaskMutation,
  EditTaskMutationVariables,
  DeleteTaskMutation,
  DeleteTaskMutationVariables
} from './../../../../client/src/graphql/types'
import {
  GQL_CREATE_TASK,
  GQL_EDIT_TASK,
  GQL_DELETE_TASK
} from './../../../../client/src/graphql/mutations/task'
import { gqlReq, projToSnapshot } from '../../testUtils'

describe('Task Mutations', () => {
  test('createTask', async () => {
    const PROJECT_ID = '5d08076bb7fc44284546db2d'
    const COLUMN_ID = '5d08076bb7fc44284546db2e'

    const { data: createData } = await gqlReq<
      CreateTaskMutation,
      CreateTaskMutationVariables
    >({
      isMutation: true,
      query: GQL_CREATE_TASK,
      variables: {
        projId: PROJECT_ID,
        columnId: COLUMN_ID,
        taskInfo: {
          name: 'new task name'
        }
      }
    })

    expect(createData).toBeTruthy()

    const task = createData!.createTask!.task!

    expect(task.name).toBe('new task name')
    expect({ ...task, id: 'TASK_ID', startDate: 'haha no' }).toMatchSnapshot()

    /* editTask */
    const { data: editData } = await gqlReq<
      EditTaskMutation,
      EditTaskMutationVariables
    >({
      isMutation: true,
      query: GQL_EDIT_TASK,
      variables: {
        projId: PROJECT_ID,
        taskId: task.id,
        newTask: {
          name: 'haha newer name'
        }
      }
    })

    expect(editData!.editTask!.task!.name).toBe('haha newer name')
    expect(editData!.editTask!.task!.id).toBe(task.id)

    expect({
      ...editData!.editTask!.task!,
      id: 'TASK_ID',
      startDate: 'noo'
    }).toMatchSnapshot()

    /* deleteTask */

    const { data: deleteData } = await gqlReq<
      DeleteTaskMutation,
      DeleteTaskMutationVariables
    >({
      isMutation: true,
      query: GQL_DELETE_TASK,
      variables: {
        projId: PROJECT_ID,
        taskId: task.id
      }
    })

    expect(deleteData!.deleteTask!.project!.tasks.length + 1).toBe(
      editData!.editTask!.project!.tasks.length
    )

    expect(projToSnapshot(deleteData!.deleteTask!.project!)).toMatchSnapshot()
  })
  // comments
  // subtasks
})
