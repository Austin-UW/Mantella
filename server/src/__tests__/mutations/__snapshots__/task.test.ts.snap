// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Task Mutations createTask 1`] = `
Object {
  "assignedUsers": Array [],
  "color": "#FFFFFF",
  "comments": Array [],
  "completed": false,
  "description": null,
  "dueDate": null,
  "id": "TASK_ID",
  "name": "new task name",
  "points": 0,
  "recurrance": null,
  "startDate": "haha no",
  "subTasks": Array [],
  "tags": Array [],
  "timeWorkedOn": 0,
}
`;

exports[`Task Mutations createTask 2`] = `
Object {
  "assignedUsers": Array [],
  "color": "#FFFFFF",
  "comments": Array [],
  "completed": false,
  "description": null,
  "dueDate": null,
  "id": "TASK_ID",
  "name": "haha newer name",
  "points": 0,
  "recurrance": null,
  "startDate": "noo",
  "subTasks": Array [],
  "tags": Array [],
  "timeWorkedOn": 0,
}
`;

exports[`Task Mutations createTask 3`] = `
Object {
  "columnIds": Array [
    "COL_ID 0",
  ],
  "columns": Array [
    Object {
      "id": "COL_ID 0",
      "isCompletedColumn": false,
      "name": "Todo",
      "taskIds": Array [
        "TASK_ID 0",
        "TASK_ID 1",
        "TASK_ID 2",
        "TASK_ID 3",
        "TASK_ID 4",
        "TASK_ID 5",
        "TASK_ID 6",
        "TASK_ID 7",
        "TASK_ID 8",
        "TASK_ID 9",
        "TASK_ID 10",
        "TASK_ID 11",
        "TASK_ID 12",
        "TASK_ID 13",
        "TASK_ID 14",
        "TASK_ID 15",
        "TASK_ID 16",
      ],
      "taskLimit": 0,
    },
  ],
  "id": "PROJECT_ID",
  "name": "PHOON_TIME",
  "ownerId": "OWNER_ID",
  "swimlanes": Array [],
  "tags": Array [],
  "tasks": Array [
    Object {
      "assignedUsers": Array [],
      "color": "#FFFFFF",
      "comments": Array [],
      "completed": false,
      "description": null,
      "dueDate": null,
      "id": "TASK_ID 0",
      "name": "new task name",
      "points": 0,
      "recurrance": null,
      "startDate": 2019-06-17T22:04:54.777Z,
      "subTasks": Array [],
      "tags": Array [],
      "timeWorkedOn": 0,
    },
    Object {
      "assignedUsers": Array [],
      "color": "#FFFFFF",
      "comments": Array [],
      "completed": false,
      "description": null,
      "dueDate": null,
      "id": "TASK_ID 1",
      "name": "new task name",
      "points": 0,
      "recurrance": null,
      "startDate": 2019-06-17T22:30:41.705Z,
      "subTasks": Array [],
      "tags": Array [],
      "timeWorkedOn": 0,
    },
    Object {
      "assignedUsers": Array [],
      "color": "#FFFFFF",
      "comments": Array [],
      "completed": false,
      "description": null,
      "dueDate": null,
      "id": "TASK_ID 2",
      "name": "new task name",
      "points": 0,
      "recurrance": null,
      "startDate": 2019-06-17T23:11:47.771Z,
      "subTasks": Array [],
      "tags": Array [],
      "timeWorkedOn": 0,
    },
    Object {
      "assignedUsers": Array [],
      "color": "#FFFFFF",
      "comments": Array [],
      "completed": false,
      "description": null,
      "dueDate": null,
      "id": "TASK_ID 3",
      "name": "new task name",
      "points": 0,
      "recurrance": null,
      "startDate": 2019-06-17T23:12:18.699Z,
      "subTasks": Array [],
      "tags": Array [],
      "timeWorkedOn": 0,
    },
    Object {
      "assignedUsers": Array [],
      "color": "#FFFFFF",
      "comments": Array [],
      "completed": false,
      "description": null,
      "dueDate": null,
      "id": "TASK_ID 4",
      "name": "new task name",
      "points": 0,
      "recurrance": null,
      "startDate": 2019-06-17T23:12:25.267Z,
      "subTasks": Array [],
      "tags": Array [],
      "timeWorkedOn": 0,
    },
    Object {
      "assignedUsers": Array [],
      "color": "#FFFFFF",
      "comments": Array [],
      "completed": false,
      "description": null,
      "dueDate": null,
      "id": "TASK_ID 5",
      "name": "new task name",
      "points": 0,
      "recurrance": null,
      "startDate": 2019-06-17T23:12:35.396Z,
      "subTasks": Array [],
      "tags": Array [],
      "timeWorkedOn": 0,
    },
    Object {
      "assignedUsers": Array [],
      "color": "#FFFFFF",
      "comments": Array [],
      "completed": false,
      "description": null,
      "dueDate": null,
      "id": "TASK_ID 6",
      "name": "new task name",
      "points": 0,
      "recurrance": null,
      "startDate": 2019-06-17T23:17:06.381Z,
      "subTasks": Array [],
      "tags": Array [],
      "timeWorkedOn": 0,
    },
    Object {
      "assignedUsers": Array [],
      "color": "#FFFFFF",
      "comments": Array [],
      "completed": false,
      "description": null,
      "dueDate": null,
      "id": "TASK_ID 7",
      "name": "new task name",
      "points": 0,
      "recurrance": null,
      "startDate": 2019-06-17T23:17:30.959Z,
      "subTasks": Array [],
      "tags": Array [],
      "timeWorkedOn": 0,
    },
    Object {
      "assignedUsers": Array [],
      "color": "#FFFFFF",
      "comments": Array [],
      "completed": false,
      "description": null,
      "dueDate": null,
      "id": "TASK_ID 8",
      "name": "haha newer name",
      "points": 0,
      "recurrance": null,
      "startDate": 2019-06-17T23:19:14.389Z,
      "subTasks": Array [],
      "tags": Array [],
      "timeWorkedOn": 0,
    },
    Object {
      "assignedUsers": Array [],
      "color": "#FFFFFF",
      "comments": Array [],
      "completed": false,
      "description": null,
      "dueDate": null,
      "id": "TASK_ID 9",
      "name": "haha newer name",
      "points": 0,
      "recurrance": null,
      "startDate": 2019-06-17T23:19:22.684Z,
      "subTasks": Array [],
      "tags": Array [],
      "timeWorkedOn": 0,
    },
    Object {
      "assignedUsers": Array [],
      "color": "#FFFFFF",
      "comments": Array [],
      "completed": false,
      "description": null,
      "dueDate": null,
      "id": "TASK_ID 10",
      "name": "haha newer name",
      "points": 0,
      "recurrance": null,
      "startDate": 2019-06-17T23:20:27.884Z,
      "subTasks": Array [],
      "tags": Array [],
      "timeWorkedOn": 0,
    },
    Object {
      "assignedUsers": Array [],
      "color": "#FFFFFF",
      "comments": Array [],
      "completed": false,
      "description": null,
      "dueDate": null,
      "id": "TASK_ID 11",
      "name": "haha newer name",
      "points": 0,
      "recurrance": null,
      "startDate": 2019-06-17T23:20:34.478Z,
      "subTasks": Array [],
      "tags": Array [],
      "timeWorkedOn": 0,
    },
    Object {
      "assignedUsers": Array [],
      "color": "#FFFFFF",
      "comments": Array [],
      "completed": false,
      "description": null,
      "dueDate": null,
      "id": "TASK_ID 12",
      "name": "haha newer name",
      "points": 0,
      "recurrance": null,
      "startDate": 2019-06-17T23:22:00.690Z,
      "subTasks": Array [],
      "tags": Array [],
      "timeWorkedOn": 0,
    },
    Object {
      "assignedUsers": Array [],
      "color": "#FFFFFF",
      "comments": Array [],
      "completed": false,
      "description": null,
      "dueDate": null,
      "id": "TASK_ID 13",
      "name": "haha newer name",
      "points": 0,
      "recurrance": null,
      "startDate": 2019-06-17T23:22:48.985Z,
      "subTasks": Array [],
      "tags": Array [],
      "timeWorkedOn": 0,
    },
    Object {
      "assignedUsers": Array [],
      "color": "#FFFFFF",
      "comments": Array [],
      "completed": false,
      "description": null,
      "dueDate": null,
      "id": "TASK_ID 14",
      "name": "haha newer name",
      "points": 0,
      "recurrance": null,
      "startDate": 2019-06-17T23:22:57.911Z,
      "subTasks": Array [],
      "tags": Array [],
      "timeWorkedOn": 0,
    },
    Object {
      "assignedUsers": Array [],
      "color": "#FFFFFF",
      "comments": Array [],
      "completed": false,
      "description": null,
      "dueDate": null,
      "id": "TASK_ID 15",
      "name": "haha newer name",
      "points": 0,
      "recurrance": null,
      "startDate": 2019-06-17T23:23:03.642Z,
      "subTasks": Array [],
      "tags": Array [],
      "timeWorkedOn": 0,
    },
    Object {
      "assignedUsers": Array [],
      "color": "#FFFFFF",
      "comments": Array [],
      "completed": false,
      "description": null,
      "dueDate": null,
      "id": "TASK_ID 16",
      "name": "haha newer name",
      "points": 0,
      "recurrance": null,
      "startDate": 2019-06-17T23:24:33.379Z,
      "subTasks": Array [],
      "tags": Array [],
      "timeWorkedOn": 0,
    },
  ],
  "users": Array [],
}
`;
